File: backend/app/utils/validation/validators.py
"""
Reusable validation utility functions.
"""

import json
import re
from datetime import date, datetime
from typing import Any
from urllib.parse import urlparse


def validate_future_date_strict(
    check_date: date | datetime,
    min_days_ahead: int = 1,
) -> tuple[bool, str | None]:
    """
    Strictly validate date is in the future.

    Args:
        check_date: Date to validate
        min_days_ahead: Minimum days ahead required

    Returns:
        tuple: (is_valid, error_message)
    """
    if isinstance(check_date, datetime):
        check_date = check_date.date()

    today = date.today()
    min_date = today

    if min_days_ahead > 0:
        from datetime import timedelta
        min_date = today + timedelta(days=min_days_ahead)

    if check_date < min_date:
        return False, f"Date must be at least {min_days_ahead} days in the future"

    return True, None


def validate_email_domain(
    email: str,
    allowed_domains: list[str] | None = None,
    blocked_domains: list[str] | None = None,
) -> tuple[bool, str | None]:
    """
    Validate email domain restrictions.

    Args:
        email: Email address to validate
        allowed_domains: List of allowed domains (None = all allowed)
        blocked_domains: List of blocked domains

    Returns:
        tuple: (is_valid, error_message)
    """
    if not email or "@" not in email:
        return False, "Invalid email format"

    domain = email.split("@")[1].lower()

    # Check blocked domains
    if blocked_domains and domain in [d.lower() for d in blocked_domains]:
        return False, f"Email domain '{domain}' is not allowed"

    # Check allowed domains
    if allowed_domains:
        if domain not in [d.lower() for d in allowed_domains]:
            return False, f"Email domain must be one of: {', '.join(allowed_domains)}"

    return True, None


def validate_url(
    url: str,
    allowed_schemes: list[str] | None = None,
    require_tld: bool = True,
) -> tuple[bool, str | None]:
    """
    Validate URL format and restrictions.

    Args:
        url: URL to validate
        allowed_schemes: List of allowed schemes (default: ['http', 'https'])
        require_tld: Require top-level domain

    Returns:
        tuple: (is_valid, error_message)
    """
    if not url:
        return False, "URL cannot be empty"

    if allowed_schemes is None:
        allowed_schemes = ["http", "https"]

    try:
        parsed = urlparse(url)

        # Check scheme
        if parsed.scheme not in allowed_schemes:
            return False, f"URL scheme must be one of: {', '.join(allowed_schemes)}"

        # Check hostname exists
        if not parsed.netloc:
            return False, "URL must have a hostname"

        # Check for TLD if required
        if require_tld:
            if "." not in parsed.netloc:
                return False, "URL must have a top-level domain"

        return True, None

    except Exception as e:
        return False, f"Invalid URL format: {str(e)}"


def validate_json(
    json_string: str,
    schema: dict | None = None,
) -> tuple[bool, str | None, dict | None]:
    """
    Validate JSON string and optionally validate against schema.

    Args:
        json_string: JSON string to validate
        schema: Optional JSON schema to validate against

    Returns:
        tuple: (is_valid, error_message, parsed_json)
    """
    if not json_string:
        return False, "JSON string cannot be empty", None

    try:
        parsed = json.loads(json_string)

        # Optional: validate against schema
        # (Would use jsonschema library here)

        return True, None, parsed

    except json.JSONDecodeError as e:
        return False, f"Invalid JSON: {str(e)}", None
    except Exception as e:
        return False, f"JSON validation error: {str(e)}", None


def validate_password_strength(
    password: str,
    min_length: int = 12,
    require_uppercase: bool = True,
    require_lowercase: bool = True,
    require_digit: bool = True,
    require_special: bool = True,
) -> tuple[bool, list[str]]:
    """
    Validate password meets strength requirements.

    Args:
        password: Password to validate
        min_length: Minimum length
        require_uppercase: Require uppercase letter
        require_lowercase: Require lowercase letter
        require_digit: Require digit
        require_special: Require special character

    Returns:
        tuple: (is_valid, list of failed requirements)
    """
    failed_requirements = []

    if len(password) < min_length:
        failed_requirements.append(f"Password must be at least {min_length} characters")

    if require_uppercase and not re.search(r"[A-Z]", password):
        failed_requirements.append("Password must contain at least one uppercase letter")

    if require_lowercase and not re.search(r"[a-z]", password):
        failed_requirements.append("Password must contain at least one lowercase letter")

    if require_digit and not re.search(r"\d", password):
        failed_requirements.append("Password must contain at least one digit")

    if require_special and not re.search(r"[!@#$%^&*(),.?\":{}|<>]", password):
        failed_requirements.append("Password must contain at least one special character")

    return len(failed_requirements) == 0, failed_requirements


def validate_file_extension(
    filename: str,
    allowed_extensions: list[str],
) -> tuple[bool, str | None]:
    """
    Validate file has allowed extension.

    Args:
        filename: Filename to validate
        allowed_extensions: List of allowed extensions (e.g., ['pdf', 'doc'])

    Returns:
        tuple: (is_valid, error_message)
    """
    if not filename:
        return False, "Filename cannot be empty"

    if "." not in filename:
        return False, "Filename must have an extension"

    extension = filename.rsplit(".", 1)[1].lower()

    allowed_lower = [ext.lower() for ext in allowed_extensions]

    if extension not in allowed_lower:
        return False, f"File extension must be one of: {', '.join(allowed_extensions)}"

    return True, None


def validate_file_size(
    size_bytes: int,
    max_size_mb: float,
) -> tuple[bool, str | None]:
    """
    Validate file size is within limits.

    Args:
        size_bytes: File size in bytes
        max_size_mb: Maximum size in megabytes

    Returns:
        tuple: (is_valid, error_message)
    """
    max_bytes = int(max_size_mb * 1024 * 1024)

    if size_bytes > max_bytes:
        actual_mb = size_bytes / (1024 * 1024)
        return False, f"File size ({actual_mb:.1f}MB) exceeds maximum ({max_size_mb}MB)"

    return True, None


def validate_list_unique(
    items: list[Any],
    field: str | None = None,
) -> tuple[bool, str | None]:
    """
    Validate list contains only unique items.

    Args:
        items: List to validate
        field: Optional field name for objects

    Returns:
        tuple: (is_valid, error_message)
    """
    if field:
        # For list of objects
        values = [getattr(item, field) if hasattr(item, field) else item.get(field) for item in items]
    else:
        # For list of primitives
        values = items

    if len(values) != len(set(values)):
        return False, "List contains duplicate items"

    return True, None

File: frontend/src/lib/validation/validators.ts
/**
 * Custom validation functions for frontend.
 */

/**
 * Validate future date.
 */
export function validateFutureDate(
  dateString: string,
  minDaysAhead: number = 0
): { isValid: boolean; error?: string } {
  const inputDate = new Date(dateString);
  const today = new Date();
  today.setHours(0, 0, 0, 0);

  const minDate = new Date(today);
  minDate.setDate(minDate.getDate() + minDaysAhead);

  if (inputDate < minDate) {
    return {
      isValid: false,
      error:
        minDaysAhead > 0
          ? `Date must be at least ${minDaysAhead} days in the future`
          : "Date must be in the future",
    };
  }

  return { isValid: true };
}

/**
 * Validate date range.
 */
export function validateDateRange(
  startDate: string,
  endDate: string
): { isValid: boolean; error?: string } {
  const start = new Date(startDate);
  const end = new Date(endDate);

  if (end < start) {
    return {
      isValid: false,
      error: "End date must be after start date",
    };
  }

  return { isValid: true };
}

/**
 * Validate email domain.
 */
export function validateEmailDomain(
  email: string,
  allowedDomains?: string[]
): { isValid: boolean; error?: string } {
  if (!email || !email.includes("@")) {
    return { isValid: false, error: "Invalid email format" };
  }

  const domain = email.split("@")[1].toLowerCase();

  if (allowedDomains && allowedDomains.length > 0) {
    const isAllowed = allowedDomains.some((d) => d.toLowerCase() === domain);
    if (!isAllowed) {
      return {
        isValid: false,
        error: `Email domain must be one of: ${allowedDomains.join(", ")}`,
      };
    }
  }

  return { isValid: true };
}

/**
 * Validate file extension.
 */
export function validateFileExtension(
  filename: string,
  allowedExtensions: string[]
): { isValid: boolean; error?: string } {
  if (!filename || !filename.includes(".")) {
    return { isValid: false, error: "File must have an extension" };
  }

  const extension = filename.split(".").pop()?.toLowerCase();

  if (!extension || !allowedExtensions.includes(extension)) {
    return {
      isValid: false,
      error: `File extension must be one of: ${allowedExtensions.join(", ")}`,
    };
  }

  return { isValid: true };
}

/**
 * Validate file size.
 */
export function validateFileSize(
  sizeBytes: number,
  maxSizeMB: number
): { isValid: boolean; error?: string } {
  const maxBytes = maxSizeMB * 1024 * 1024;

  if (sizeBytes > maxBytes) {
    const actualMB = (sizeBytes / (1024 * 1024)).toFixed(1);
    return {
      isValid: false,
      error: `File size (${actualMB}MB) exceeds maximum (${maxSizeMB}MB)`,
    };
  }

  return { isValid: true };
}

/**
 * Validate PGY level for person type.
 */
export function validatePgyLevel(
  pgyLevel: number | null,
  personType: string
): { isValid: boolean; error?: string } {
  if (personType === "faculty") {
    if (pgyLevel !== null) {
      return {
        isValid: false,
        error: "Faculty cannot have a PGY level",
      };
    }
  } else if (personType === "resident") {
    if (pgyLevel === null) {
      return {
        isValid: false,
        error: "Residents must have a PGY level",
      };
    }
    if (pgyLevel < 1 || pgyLevel > 3) {
      return {
        isValid: false,
        error: "PGY level must be between 1 and 3",
      };
    }
  }

  return { isValid: true };
}

/**
 * Validate faculty role for person type.
 */
export function validateFacultyRole(
  facultyRole: string | null,
  personType: string
): { isValid: boolean; error?: string } {
  if (personType === "resident" && facultyRole !== null) {
    return {
      isValid: false,
      error: "Residents cannot have a faculty role",
    };
  }

  return { isValid: true };
}

/**
 * Validate assignment role compatibility.
 */
export function validateAssignmentRole(
  role: string,
  personType: string
): { isValid: boolean; error?: string } {
  if (role === "supervising" && personType !== "faculty") {
    return {
      isValid: false,
      error: "Only faculty can have 'supervising' role",
    };
  }

  return { isValid: true };
}

/**
 * Validate unique values in list.
 */
export function validateUniqueValues<T>(
  values: T[],
  fieldName?: string
): { isValid: boolean; error?: string; duplicates?: T[] } {
  const seen = new Set<T>();
  const duplicates: T[] = [];

  for (const value of values) {
    if (seen.has(value)) {
      duplicates.push(value);
    } else {
      seen.add(value);
    }
  }

  if (duplicates.length > 0) {
    return {
      isValid: false,
      error: fieldName
        ? `${fieldName} contains duplicate values`
        : "List contains duplicate values",
      duplicates,
    };
  }

  return { isValid: true };
}

/**
 * Validate password strength.
 */
export function validatePasswordStrength(password: string): {
  isValid: boolean;
  errors: string[];
  strength: "weak" | "medium" | "strong";
} {
  const errors: string[] = [];
  let score = 0;

  if (password.length < 12) {
    errors.push("Password must be at least 12 characters");
  } else {
    score += 1;
  }

  if (!/[A-Z]/.test(password)) {
    errors.push("Password must contain at least one uppercase letter");
  } else {
    score += 1;
  }

  if (!/[a-z]/.test(password)) {
    errors.push("Password must contain at least one lowercase letter");
  } else {
    score += 1;
  }

  if (!/\d/.test(password)) {
    errors.push("Password must contain at least one digit");
  } else {
    score += 1;
  }

  if (!/[!@#$%^&*(),.?":{}|<>]/.test(password)) {
    errors.push("Password must contain at least one special character");
  } else {
    score += 1;
  }

  // Additional strength checks
  if (password.length >= 16) score += 1;
  if (/[!@#$%^&*(),.?":{}|<>]{2,}/.test(password)) score += 1;

  let strength: "weak" | "medium" | "strong";
  if (score < 3) strength = "weak";
  else if (score < 5) strength = "medium";
  else strength = "strong";

  return {
    isValid: errors.length === 0,
    errors,
    strength,
  };
}
